### PR TITLE
Plugin manager rewrite as plugin installers

### PR DESCRIPTION
**Pull request by skrzypo987**
## Description
This PR rewrites the plugin installation process into smaller installers using mechanism from https://github.com/trinodb/trino/pull/16060. That way the logic of installing plugins i s moved into classes, where they are actually installed, and only guice bindings link the two together.

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/server/ServerMainModule.java [hunk 1]
```diff
@@ -439,7 +445,15 @@ protected void setup(Binder binder)
         newOptionalBinder(binder, PluginsProvider.class).setDefault()
                 .to(ServerPluginsProvider.class).in(Scopes.SINGLETON);
         configBinder(binder).bindConfig(ServerPluginsProviderConfig.class);
-        newSetBinder(binder, PluginInstaller.class);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/ServerMainModule.java]: extract variable for multi set binder.

File: core/trino-main/src/main/java/io/trino/testing/LocalQueryRunner.java [hunk 2]
```diff
@@ -469,19 +461,8 @@ private LocalQueryRunner(
...
-                new CertificateAuthenticatorManager(),
-                Optional.of(new HeaderAuthenticatorManager(new HeaderAuthenticatorConfig())),
-                eventListenerManager,
-                new GroupProviderManager(),
-                new SessionPropertyDefaults(nodeInfo, accessControl),
                 typeRegistry,
-                blockEncodingManager,
                 handleResolver,
-                exchangeManagerRegistry,
                 ImmutableSet.of());
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/testing/LocalQueryRunner.java:466]: I don't think it works now as you need to pass all installers here.

File: core/trino-main/src/main/java/io/trino/exchange/ExchangeManagerRegistry.java [hunk 3]
```diff
@@ -47,6 +50,15 @@
 
     private volatile ExchangeManager exchangeManager;
 
+    @Override
+    public void installPlugin(Plugin plugin)
+    {
+        for (ExchangeManagerFactory exchangeManagerFactory : plugin.getExchangeManagerFactories()) {
+            log.info("Registering exchange manager %s", exchangeManagerFactory.getName());
```
[the remark this answers is not in the document]
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/exchange/ExchangeManagerRegistry.java:57]: A simple example
```
presto-master       | 2023-02-14T13:45:46.700+0545      INFO    main    io.trino.server.PluginManager   -- Loading plugin /var/presto/plugin/exchange-filesystem --
presto-master       | 2023-02-14T13:45:46.720+0545      INFO    main    io.trino.server.PluginManager   Installing io.trino.plugin.exchange.filesystem.FileSystemExchangePlugin
presto-master       | 2023-02-14T13:45:46.727+0545      INFO    main    io.trino.server.PluginManager   Registering exchange manager filesystem
presto-master       | 2023-02-14T13:45:46.728+0545      INFO    main    io.trino.server.PluginManager   -- Finished loading plugin /var/presto/plugin/exchange-filesystem --

presto-master       | 2023-02-14T14:16:08.464+0545      INFO    main    io.trino.server.PluginManager   -- Loading plugin /var/presto/plugin/exchange-filesystem --
presto-master       | 2023-02-14T14:16:08.489+0545      INFO    main    io.trino.server.PluginManager   Installing io.trino.plugin.exchange.filesystem.FileSystemExchangePlugin
presto-master       | 2023-02-14T14:16:08.499+0545      INFO    main    io.trino.exchange.ExchangeManagerRegistry       Registering exchange manager filesystem
presto-master       | 2023-02-14T14:16:08.500+0545      INFO    main    io.trino.server.PluginManager   -- Finished loading plugin /var/presto/plugin/exchange-filesystem --

```

**Comment by skrzypo987** [comment]
@kokosing AC

**Review by kokosing — APPROVED** [review]

**Comment by kokosing** [comment]
Please rebase

**Review by ksobolew — APPROVED** [review]
Just wanted to say that I like it

**Review by martint — CHANGES_REQUESTED** [review]
Classes such as ResourceGroupManager, TypeRegistry, GlobalFunctionCatalog, etc., should not have to be aware of plugin infrastructure -- it should be the opposite.

**Comment by kokosing** [comment]
In such case as I understand you, that means that we should have dedicated plugin installers per such entity. For example `ResourceGroupManagerPluginInstaller` that would just glue the plugin infrastructure with `ResourceGroupManager`.

**Pull request closed without merging**